services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/functionaltests/AsyncIndexScanTest.java

```diff
@@ -0,0 +1,290 @@
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(id, sort, value, gsiId, gsiSort);
+        }
+    }
+
+    private static final TableSchema<Record> TABLE_SCHEMA =
+        StaticTableSchema.builder(Record.class)
+                         .newItemSupplier(Record::new)
+                         .attributes(
+                             stringAttribute("id", Record::getId, Record::setId).as(primaryPartitionKey()),
+                             integerNumberAttribute("sort", Record::getSort, Record::setSort).as(primarySortKey()),
+                             integerNumberAttribute("value", Record::getValue, Record::setValue),
+                             stringAttribute("gsi_id", Record::getGsiId, Record::setGsiId)
+                                 .as(secondaryPartitionKey("gsi_keys_only")),
+                             integerNumberAttribute("gsi_sort", Record::getGsiSort, Record::setGsiSort)
+                                 .as(secondarySortKey("gsi_keys_only")))
+                         .build();
+
+    private static final List<Record> RECORDS =
+        IntStream.range(0, 10)
+                 .mapToObj(i -> new Record()
+                     .setId("id-value")
+                     .setSort(i)
+                     .setValue(i)
+                     .setGsiId("gsi-id-value")
+                     .setGsiSort(i))
+                 .collect(Collectors.toList());
+
+    private static final List<Record> KEYS_ONLY_RECORDS =
+        RECORDS.stream()
+               .map(record -> new Record()
+                   .setId(record.id)
+                   .setSort(record.sort)
+                   .setGsiId(record.gsiId)
+                   .setGsiSort(record.gsiSort))
+               .collect(Collectors.toList());
+
+    private DynamoDbEnhancedAsyncClient enhancedAsyncClient =
+        DefaultDynamoDbEnhancedAsyncClient.builder()
+                                          .dynamoDbClient(getDynamoDbAsyncClient())
+                                          .build();
+
+    private DynamoDbAsyncTable<Record> mappedTable = enhancedAsyncClient.table(getConcreteTableName("table-name"), TABLE_SCHEMA);
+    private DynamoDbAsyncIndex<Record> keysOnlyMappedIndex = mappedTable.index("gsi_keys_only");
+
+    private void insertRecords() {
+        RECORDS.forEach(record -> mappedTable.putItem(Record.class, r -> r.item(record)).join());
+    }
+
+    private static <T> List<T> drainPublisher(SdkPublisher<T> publisher, int expectedNumberOfResults) {
+        BufferingSubscriber<T> subscriber = new BufferingSubscriber<>();
+        publisher.subscribe(subscriber);
+        subscriber.waitForCompletion(1000L);
+
+        assertThat(subscriber.isCompleted(), is(true));
+        assertThat(subscriber.bufferedError(), is(nullValue()));
+        assertThat(subscriber.bufferedItems().size(), is(expectedNumberOfResults));
+
+        return subscriber.bufferedItems();
+    }
```

Comment:
I think we're using this code in a bunch of places, can we refactor it into a base test class or util?